polaris-react/src/components/AlphaFilters/AlphaFilters.tsx

```diff
@@ -160,28 +163,53 @@ export function AlphaFilters({
     ...pinnedFiltersFromLocalState,
   ];
 
-  const additionalFilters = filters.reduce<ActionListItemProps[]>(
-    (acc, filter) =>
-      !pinnedFilters.some(({key}) => key === filter.key)
-        ? (acc.push({
-            ...filter,
-            content: filter.label,
-            onAction: () => {
-              // PopoverOverlay will cause a rerender of the component and nuke the
-              // popoverActive state, so we set this as a microtask
-              setTimeout(() => {
-                setLocalPinnedFilters((currentLocalPinnedFilters) => [
-                  ...new Set([...currentLocalPinnedFilters, filter.key]),
-                ]);
-                filter.onAction?.();
-                togglePopoverActive();
-              }, 0);
-            },
-          }),
-          acc)
-        : acc,
-    [],
-  );
+  const onFilterClick =
+    ({key, onAction}: FilterInterface) =>
+    () => {
+      // PopoverOverlay will cause a rerender of the component and nuke the
+      // popoverActive state, so we set this as a microtask
+      setTimeout(() => {
+        setLocalPinnedFilters((currentLocalPinnedFilters) => [
+          ...new Set([...currentLocalPinnedFilters, key]),
+        ]);
+        onAction?.();
+        togglePopoverActive();
+      }, 0);
+    };
+
+  const unsectionedFilters: ActionListItemDescriptor[] = [];
+  const sectionedFilters: {
+    title: string;
+    items: ActionListItemDescriptor[];
+  }[] = [];
+  const sectionedFilterKeyIndex: {
+    [key: string]: number;
+  } = {};
+
+  filters.forEach((filter) => {
+    if (pinnedFilters.some(({key}) => key === filter.key)) return;
+
+    const filterActionItem = {
+      ...filter,
+      content: filter.label,
+      onAction: onFilterClick(filter),
+    };
+
+    if (filter.section) {
+      if (sectionedFilterKeyIndex[filter.section] === undefined) {
+        sectionedFilterKeyIndex[filter.section] = sectionedFilters.length;
+        sectionedFilters.push({
+          title: filter.section,
+          items: [filterActionItem],
+        });
+      } else {
+        const sectionIndex = sectionedFilterKeyIndex[filter.section];
+        sectionedFilters[sectionIndex].items.push(filterActionItem);
+      }
+    } else {
+      unsectionedFilters.push(filterActionItem);
+    }
+  });
```

Comment:
You know what, I was thinking about this all wrong. I was thinking the `findIndex` needs to search through all of the items but it's only searching through the array of sections. Even if we have like 5 filter sections (extremely unlikely) that's not an expensive operation. So ya let's do it as you've suggested.

Reply:
Totally aligned. Given the small number of items, it seemed reasonable to omit the index map object. If we had lots of filters to process, it would be best for performance to create that separately 👍

I'm happy with either option whichever feels most maintainable, thank you for exploring this more!